### PR TITLE
Enable rpm/deb build on dev machines

### PR DESCRIPTION
**Pull request by NikolaMilosavljevic**
Enable building DEB and RPM packages with a new switch:

./build.sh /p:BuildLinuxPackages=true

### CODE REVIEW

File: src/installer/pkg/packaging/rpm/package.targets [hunk 1]
```diff
@@ -315,84 +318,31 @@
...
-      <RpmIntermediatesDir>$(PackagesIntermediateDir)$(RpmPackageName)/$(RpmPackageVersion)</RpmIntermediatesDir>
-    </PropertyGroup>
+  <Target Name="GenerateRuntimeDependenciesRpms">
 
     <PropertyGroup>
-      <RpmLayoutDirectory>$(RpmIntermediatesDir)/rpmLayoutDirectory/</RpmLayoutDirectory>
-      <RpmLayoutPackageRoot>$(RpmLayoutDirectory)package_root</RpmLayoutPackageRoot>
-      <RpmLayoutDocs>$(RpmLayoutDirectory)docs</RpmLayoutDocs>
-      <RpmLayoutTemplates>$(RpmLayoutDirectory)templates</RpmLayoutTemplates> <!-- Copyright, Changelog -->
+      <SharedProperties>rpmPackagingConfigPath=$(rpmPackagingConfigPath);DotnetRuntimeDependenciesPackageInstallerFile=$(DotnetRuntimeDependenciesPackageInstallerFile);OutputRid=$(OutputRid);RuntimeDependenciesRpmPkgName=$(RuntimeDependenciesRpmPkgName);RuntimeDepsRpmPackageVersion=$(RuntimeDepsRpmPackageVersion);RuntimeDepsRpmPackageRelease=$(RuntimeDepsRpmPackageRelease);TemplatesDir=$(TemplatesDir)</SharedProperties>
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/package.targets]: Please split this onto multiple lines--I think newlines are ok, but if not, repeatedly appending to the prop is a typical way to do it.

File: src/installer/pkg/packaging/rpm/package.targets [hunk 2]
```diff
@@ -315,84 +318,31 @@
...
-        <ReplacementString>$(RuntimeDependenciesRpmPkgName)</ReplacementString>
-      </RuntimeDependenciesTokenValue>
-      <RuntimeDependenciesTokenValue Include="%RUNTIME_DEPS_VERSION%">
-        <ReplacementString>$(RuntimeDepsRpmPackageVersion)</ReplacementString> 
-      </RuntimeDependenciesTokenValue>
-      <RuntimeDependenciesTokenValue Include="%RUNTIME_DEPS_REVISION%">
-        <ReplacementString>$(RuntimeDepsRpmPackageRelease)</ReplacementString>
-      </RuntimeDependenciesTokenValue>
+    <!-- non-portable build - add OutputRid to new item list -->
+    <ItemGroup Condition="'$(PortableBuild)' == 'false'">
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/package.targets]: `!= 'true'` is the usual idiom, it handles the case where the property isn't set at all (or set to some unusual value).

File: src/installer/pkg/packaging/rpm/package.targets [hunk 3]
```diff
@@ -315,84 +318,31 @@
...
-      <RuntimeDependenciesTokenValue Include="%RUNTIME_DEPS_RPM_PACKAGE_NAME%">
-        <ReplacementString>$(RuntimeDependenciesRpmPkgName)</ReplacementString>
-      </RuntimeDependenciesTokenValue>
-      <RuntimeDependenciesTokenValue Include="%RUNTIME_DEPS_VERSION%">
-        <ReplacementString>$(RuntimeDepsRpmPackageVersion)</ReplacementString> 
-      </RuntimeDependenciesTokenValue>
-      <RuntimeDependenciesTokenValue Include="%RUNTIME_DEPS_REVISION%">
-        <ReplacementString>$(RuntimeDepsRpmPackageRelease)</ReplacementString>
-      </RuntimeDependenciesTokenValue>
+    <!-- non-portable build - add OutputRid to new item list -->
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/package.targets]: I'd prefer describing the purpose here (and similar in below comments):
```suggestion
    <!-- non-portable build - build a single dotnet-runtime-deps package using OutputRid -->
```

File: src/installer/pkg/packaging/rpm/package.targets [hunk 4]
```diff
@@ -315,84 +318,31 @@
...
-
-    <!-- Copy package to output -->
-    <ItemGroup>
-      <GeneratedRpmFiles Remove="@(GeneratedRpmFiles)" />
-      <GeneratedRpmFiles Include="$(RpmIntermediatesDir)/*.rpm" />
+    <!-- portable build - find all config json deps files and add their OutputRid values to new item list -->
+    <!-- iterate thru all items in item list and call pack-deps project -->
+    <ItemGroup Condition="'$(PortableBuild)' == 'true'">
+      <ConfigJsonFilenames Include="$(rpmPackagingConfigPath)$(DepsConfigFilePrefix)*.json" />
+      <Rids Include="$([System.Text.RegularExpressions.Regex]::Match(%(ConfigJsonFilenames.Filename),'$(DepsConfigFilePrefix)([A-Za-z0-9_\-\.]+)').Groups[1].Value)" />
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/package.targets]: A little bit of formatting--quotes and space after `,`.

```suggestion
      <Rids Include="$([System.Text.RegularExpressions.Regex]::Match('%(ConfigJsonFilenames.Filename)', '$(DepsConfigFilePrefix)([A-Za-z0-9_\-\.]+)').Groups[1].Value)" />
```

File: src/installer/pkg/packaging/rpm/package.targets [hunk 5]
```diff
@@ -315,84 +318,31 @@
...
     </ItemGroup>
 
-    <Error Text="@(GeneratedRpmFiles->Count()) .rpm files generated." Condition="'@(GeneratedRpmFiles->Count())' != 1" />
-
-    <Copy SourceFiles="@(GeneratedRpmFiles)"
-          DestinationFiles="$(RpmFile)"
-          OverwriteReadOnlyFiles="True"
-          SkipUnchangedFiles="False"
-          UseHardlinksIfPossible="False" />
+      <MSBuild Projects="@(DepsProjects)" Targets="GenerateSingleRuntimeDependenciesRpm" />
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/package.targets]: Looks like this is indented too far.

File: src/installer/pkg/packaging/rpm/pack-deps.proj [hunk 6]
```diff
@@ -0,0 +1,97 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/pack-deps.proj]: Can just use `<Project>`, extra stuff up here is legacy now.

File: src/installer/pkg/packaging/rpm/pack-deps.proj [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+  <Import Project='package.props' />
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/pack-deps.proj]: Double quotes are conventional

```suggestion
  <Import Project="package.props" />
```

File: src/installer/pkg/packaging/rpm/pack-deps.proj [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+  <UsingTask TaskName="ReplaceFileContents" AssemblyFile="$(InstallerTasksAssemblyPath)"/>
+  <UsingTask TaskName="BuildFPMToolPreReqs" AssemblyFile="$(InstallerTasksAssemblyPath)"/>
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/pack-deps.proj]: Missing space for style

```suggestion
  <UsingTask TaskName="BuildFPMToolPreReqs" AssemblyFile="$(InstallerTasksAssemblyPath)" />
```

File: src/installer/pkg/packaging/rpm/pack-deps.proj [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>$(NETCoreAppFramework)</TargetFramework>
+  </PropertyGroup>
+
+  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/pack-deps.proj]: Why is this import necessary?

File: src/installer/pkg/packaging/rpm/pack-deps.proj [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+  <Target Name="GenerateSingleRuntimeDependenciesRpm">
```
Comment (dagood) [inline src/installer/pkg/packaging/rpm/pack-deps.proj]: Could you leave this target where it was in `package.targets` and have the `DepsProjects` items use `$(MSBuildProjectFullPath)`? Would reduce the diff between the RPM vs. Deb targets, I think.

File: src/installer/pkg/packaging/osx/package.targets [hunk 11]
```diff
@@ -13,6 +13,11 @@
   <Target Name="GeneratePkgs"
           Condition="'$(OSGroup)' == 'OSX'"
           DependsOnTargets="GetInstallerBrandingNames;InitPkg">
+    <PropertyGroup>
+        <SharedFrameworkInstallerFile Condition="!$(SharedFrameworkInstallerFile.EndsWith('.pkg'))">$(SharedFrameworkInstallerFile).pkg</SharedFrameworkInstallerFile>
+        <SharedHostInstallerFile Condition="!$(SharedHostInstallerFile.EndsWith('.pkg'))">$(SharedHostInstallerFile).pkg</SharedHostInstallerFile>
+        <HostFxrInstallerFile Condition="!$(HostFxrInstallerFile.EndsWith('.pkg'))">$(HostFxrInstallerFile).pkg</HostFxrInstallerFile>
```
[the remark this answers is not in the document]
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/osx/package.targets]: Will fix - it seems VSCode messed this one.

File: src/installer/pkg/packaging/deb/package.targets [hunk 12]
```diff
@@ -32,6 +32,7 @@
       <DebPackageVersion>$(HostPackageVersion)</DebPackageVersion>
       <InputRoot>$(SharedHostPublishRoot)</InputRoot>
       <DebFile>$(SharedHostInstallerFile)</DebFile>
+      <DebFile Condition="!$(DebFile.EndsWith('.deb'))">$(DebFile).deb</DebFile>
```
[the remark this answers is not in the document]
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/deb/package.targets]: Great - thank you!

File: src/installer/pkg/packaging/rpm/package.targets (hunk 1 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/package.targets]: Will do.

File: src/installer/pkg/packaging/rpm/package.targets (hunk 2 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/package.targets]: Absolutely - this was missed, will fix.

File: src/installer/pkg/packaging/rpm/package.targets (hunk 3 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/package.targets]: These were meant to be to-do comments and removed - but I'll update them. 👍

File: src/installer/pkg/packaging/rpm/package.targets (hunk 5 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/package.targets]: Will fix.

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 6 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: Will fix.

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 7 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: I've only seen usage of single quotes in related source files like: https://github.com/dotnet/runtime/blob/master/src/installer/pkg/packaging/rpm/package.targets

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 8 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: Sure - I can fix this - however it was a straight copy from https://github.com/dotnet/runtime/blob/master/src/installer/pkg/packaging/rpm/package.targets

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 9 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: I don't think it is - will remove.

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 10 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: Will check the feasibility.

File: src/installer/pkg/packaging/osx/package.targets (hunk 11 above)
[the remark this answers is not in the document]
Comment (dagood) [reply src/installer/pkg/packaging/osx/package.targets]: I have this in my settings to prevent that. 🙂 

```
    "files.associations": {
        "*.msbuild": "xml",
        "*.depproj": "xml",
        "*.pkgproj": "xml",
        "*.sfxproj": "xml",
        "*.bundleproj": "xml",
        "*.builds": "xml",
        "*.proj": "xml",
        "*.props": "xml",
        "*.targets": "xml",
...
    },
```

```
    "[xml]": {
        "editor.tabSize": 2
    },
```

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 7 above)
Comment (dagood) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: Interesting, never noticed. I guess at the end of the day it's just XML... fixing over to convention would be nice though.

File: src/installer/pkg/packaging/osx/package.targets (hunk 11 above)
[the remark this answers is not in the document]
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/osx/package.targets]: Indents match the rest of this file, will keep this as is.

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 10 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: Fixing this in new commit.

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 9 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: Removing this whole proj file.

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 8 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: Removing this whole proj file.

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 7 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: Removing this whole proj file.

File: src/installer/pkg/packaging/rpm/pack-deps.proj (hunk 6 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/pack-deps.proj]: Removing this whole proj file.

File: src/installer/pkg/packaging/rpm/package.targets (hunk 4 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/packaging/rpm/package.targets]: Will fix.

**Review by dagood — COMMENTED** [review]
Looking good, a few more suggestions.

File: src/installer/pkg/Directory.Build.props [hunk 13]
```diff
@@ -36,15 +36,19 @@
     <DebtoolAssemblyFile>$(DebToolBinDir)$(Configuration)\$(NetCoreAppCurrent)\dotnet-deb-tool.dll</DebtoolAssemblyFile>
     <DotnetDebToolSourceDir>$(RepoRoot)tools-local/dotnet-deb-tool/</DotnetDebToolSourceDir>
 
+    <BuildLinuxPackages Condition="'$(BuildLinuxPackages)' != true">false</BuildLinuxPackages>
```
[the remark this answers is not in the document]
Comment (NikolaMilosavljevic) [reply src/installer/pkg/Directory.Build.props]: Sure, let me update this.

File: src/installer/Directory.Build.props [hunk 14]
```diff
@@ -304,6 +306,28 @@
         <TargetsLinux>true</TargetsLinux>
         <TargetsUnix>true</TargetsUnix>
       </PropertyGroup>
+      <Choose>
+        <When Condition="$(HostMachineRid.StartsWith('debian'))">
+          <PropertyGroup>
+            <PortableBuildDebianHost>true</PortableBuildDebianHost>
```
[the remark this answers is not in the document]
Comment (NikolaMilosavljevic) [reply src/installer/Directory.Build.props]: Will fix - thanks!

**Review by dagood — APPROVED** [review]
Thanks! 🎉
